### PR TITLE
osd: fix pg ref leaks when osd shutdown

### PR DESCRIPTION
**Pull request by yanghonggang**
Fixes: http://tracker.ceph.com/issues/20684
Signed-off-by: Yang Honggang <joseph.yang@xtaotech.com>

### CODE REVIEW

File: src/osd/PrimaryLogPG.cc [hunk 1]
```diff
@@ -250,16 +250,19 @@ class PrimaryLogPG::C_OSD_AppliedRecoveredObjectReplica : public Context {
 };
 
 // OpContext
-void PrimaryLogPG::OpContext::start_async_reads(PrimaryLogPG *pg)
```
Comment (jdurgin) [inline src/osd/PrimaryLogPG.cc]: looks like cct is unused, otherwise looks good!

File: src/osd/PrimaryLogPG.cc [hunk 2]
```diff
@@ -250,16 +250,19 @@ class PrimaryLogPG::C_OSD_AppliedRecoveredObjectReplica : public Context {
...
+  generic_dout(20) << __func__
+          << " OnReadComplete " << porc
+          << " OpContext " << this
+          << dendl;
   pg->pgbackend->objects_read_async(
-    obc->obs.oi.soid,
-    in,
-    new OnReadComplete(pg, this), pg->get_pool().fast_read);
+    obc->obs.oi.soid, in, porc, pg->get_pool().fast_read);
 }
```
Comment (liu-chunmei) [inline src/osd/PrimaryLogPG.cc]: cct is not used in this function. How to solve the problem?

File: src/osd/PrimaryLogPG.cc (hunk 1 above)
Comment (yanghonggang) [inline src/osd/PrimaryLogPG.cc]: @jdurgin  generic_dout need cct

File: src/osd/PrimaryLogPG.cc (hunk 2 above)
Comment (yanghonggang) [inline src/osd/PrimaryLogPG.cc]: @liu-chunmei  generic_dout need cct

File: src/osd/PrimaryLogPG.cc [hunk 3]
```diff
@@ -250,16 +250,19 @@ class PrimaryLogPG::C_OSD_AppliedRecoveredObjectReplica : public Context {
...
 // OpContext
-void PrimaryLogPG::OpContext::start_async_reads(PrimaryLogPG *pg)
+void PrimaryLogPG::OpContext::start_async_reads(PrimaryLogPG *pg, CephContext* cct)
 {
   inflightreads = 1;
   list<pair<boost::tuple<uint64_t, uint64_t, unsigned>,
 	    pair<bufferlist*, Context*> > > in;
   in.swap(pending_async_reads);
+  struct OnReadComplete* porc = new OnReadComplete(pg, this);
+  generic_dout(20) << __func__
```
Comment (tchaikov) [inline src/osd/PrimaryLogPG.cc]: i'd suggest revert this change. it's handy for debugging this issue, but it won't be that helpful after done with it.

**Comment by yanghonggang** [comment]
@tchaikov  ok, revert done.

File: src/osd/PrimaryLogPG.cc [hunk 4]
```diff
@@ -10651,6 +10651,19 @@ void PrimaryLogPG::on_removal(ObjectStore::Transaction *t)
     on_shutdown();
 }
 
+// clear ctxs in in_progress_async_reads
```
Comment (tchaikov) [inline src/osd/PrimaryLogPG.cc]: please remove this comment as the code is self-documenting.

**Comment by tchaikov** [comment]
aside from the nit, lgtm.

**Comment by yanghonggang** [comment]
@tchaikov comment is deleted.

**Comment by yuriw** [comment]
http://pulpito.front.sepia.ceph.com:80/yuriw-2017-07-20_19:48:38-rados-wip-yuri-testing3_2017_7_21-distro-basic-smithi/
